### PR TITLE
fix(erc4626): remove call to max functions

### LINKED ISSUE #378: call to `maxDeposit` could be avoided
on any erc4626 deposit, the bundlers calls the maxdeposit function of the 4626.

https://github.com/morpho-org/morpho-blue-bundlers/blob/23c6215801558fcedc0556aea6fc92b85c3ba8f2/src/ERC4626Bundler.sol#L63

I don't think that it's useful, depositing the maximum depositable is super super niche to my knowledge. But it costs a lot: for example for a mm deposit, we waste ~50k gas. I'm in favor of removing it.

### PR DESCRIPTION
**Pull request by Rubilmax**
- Fixes https://github.com/morpho-org/morpho-blue-bundlers-private/issues/3
- Fixes #378

### CODE REVIEW

**Review by MerlinEgalite — COMMENTED** [review]
we don't apply @QGarchery's suggestion?

**Review by Jean-Grimal — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Comment by Rubilmax** [comment]
Alternative: we could use `maxMint` (resp. `maxWithdraw`) in `mint` (resp. `withdraw`) and expect the bundle to use `deposit` and `redeem` when one wants to interact with the vault for cheap but use `mint` and `withdraw` when one wants to guarantee no revert due to slippage / rounding

@julien-devatom wdyt? We can take a call

**Comment by MerlinEgalite** [comment]
> Alternative: we could use `maxMint` (resp. `maxWithdraw`) in `mint` (resp. `withdraw`) and expect the bundle to use `deposit` and `redeem` when one wants to interact with the vault for cheap but use `mint` and `withdraw` when one wants to guarantee no revert due to slippage / rounding
> 
> @julien-devatom wdyt? We can take a call

It seems to me that it complexifies things, no?

**Comment by Rubilmax** [comment]
> It seems to me that it complexifies things, no?

In any case, the `mint` and `withdraw` function need to have a distinct behavior from `deposit` & `redeem` by nature
So it does complexify the behavior from the state of the code in this PR, but it still simplifies the complexity of the state of the code prior to this PR

And all of this is to make sure we cover broad enough usecases offchain

**Comment by Rubilmax** [comment]
Just discussed with @julien-devatom:

1. `maxMint` & `maxWithdraw` do not correspond to real offchain usecases because you don't receive what you expect (for example when bad debt is realized)
2. Using `convertToShares` (resp. `convertToAssets`) may be useful to limit to the bundler's balance, but `mint` and `withdraw` are not expected to be used offchain anyway, so it is our decision to make

**Review by adhusson — APPROVED** [review]

**Comment by MerlinEgalite** [comment]
> maxMint & maxWithdraw do not correspond to real offchain usecases because you don't receive what you expect (for example when bad debt is realized)

I don't get this. Isn't what we're suppose to mitigate ("you don't receive what you expect")?

**Review by MerlinEgalite — COMMENTED** [review]
Code seems correct. I'm ok with point 2 but need to think more about it

**Comment by Rubilmax** [comment]
> > maxMint & maxWithdraw do not correspond to real offchain usecases because you don't receive what you expect (for example when bad debt is realized)
> 
> I don't get this. Isn't what we're suppose to mitigate ("you don't receive what you expect")?

I think the only things we want to do here are:
1. make calls cheaper with MM (in `deposit` and `redeem` which are the only functions used by the frontend according to Julien)
2. make the code consistent in the behaviors across all 4 functions

1) is easily achieved, but we still want to cap to the bundler's balance of assets/shares to not revert due to prior roundings
2) is achieved by capping to what the bundler can mint/withdraw, so the bundler does not revert due to prior roundings on shares/assets acquired

**Review by Jean-Grimal — APPROVED** [review]

File: src/ERC4626Bundler.sol [hunk 1]
```diff
@@ -89,9 +85,6 @@ abstract contract ERC4626Bundler is BaseBundler {
         /// Do not check `receiver != address(this)` to allow the bundler to receive the underlying asset.
         require(receiver != address(0), ErrorsLib.ZERO_ADDRESS);
         require(owner == address(this) || owner == initiator(), ErrorsLib.UNEXPECTED_OWNER);
-
-        assets = Math.min(assets, IERC4626(vault).maxWithdraw(owner));
-
         require(assets != 0, ErrorsLib.ZERO_AMOUNT);
```
Comment (MathisGD) [inline src/ERC4626Bundler.sol:88]: not mandatory as morpho reverts on zero amount

File: src/ERC4626Bundler.sol [hunk 2]
```diff
@@ -116,7 +109,7 @@ abstract contract ERC4626Bundler is BaseBundler {
         require(receiver != address(0), ErrorsLib.ZERO_ADDRESS);
         require(owner == address(this) || owner == initiator(), ErrorsLib.UNEXPECTED_OWNER);
 
-        shares = Math.min(shares, IERC4626(vault).maxRedeem(owner));
+        shares = Math.min(shares, IERC4626(vault).balanceOf(owner));
 
         require(shares != 0, ErrorsLib.ZERO_SHARES);
```
Comment (MathisGD) [inline src/ERC4626Bundler.sol:114]: not mandatory as morpho reverts on zero amount

File: src/ERC4626Bundler.sol [hunk 3]
```diff
@@ -47,20 +44,19 @@ abstract contract ERC4626Bundler is BaseBundler {
...
-        require(receiver != address(0), ErrorsLib.ZERO_ADDRESS);
         /// Do not check `receiver != address(this)` to allow the bundler to receive the vault's shares.
+        require(receiver != address(0), ErrorsLib.ZERO_ADDRESS);
 
         address asset = IERC4626(vault).asset();
 
-        assets = Math.min(assets, IERC4626(vault).maxDeposit(receiver));
         assets = Math.min(assets, ERC20(asset).balanceOf(address(this)));
 
         require(assets != 0, ErrorsLib.ZERO_AMOUNT);
```
Comment (MathisGD) [inline src/ERC4626Bundler.sol:62]: not mandatory as morpho reverts on zero amount

File: src/ERC4626Bundler.sol [hunk 4]
```diff
@@ -22,19 +22,16 @@ abstract contract ERC4626Bundler is BaseBundler {
...
         payable
         protected
     {
-        require(receiver != address(0), ErrorsLib.ZERO_ADDRESS);
         /// Do not check `receiver != address(this)` to allow the bundler to receive the vault's shares.
-
-        shares = Math.min(shares, IERC4626(vault).maxMint(receiver));
-
+        require(receiver != address(0), ErrorsLib.ZERO_ADDRESS);
         require(shares != 0, ErrorsLib.ZERO_SHARES);
```
Comment (MathisGD) [inline src/ERC4626Bundler.sol:35]: not mandatory as morpho reverts on zero amount

**Review by MathisGD — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Review by QGarchery — APPROVED** [review]

File: src/ERC4626Bundler.sol [hunk 5]
```diff
@@ -116,7 +109,7 @@ abstract contract ERC4626Bundler is BaseBundler {
         require(receiver != address(0), ErrorsLib.ZERO_ADDRESS);
         require(owner == address(this) || owner == initiator(), ErrorsLib.UNEXPECTED_OWNER);
 
-        shares = Math.min(shares, IERC4626(vault).maxRedeem(owner));
+        shares = Math.min(shares, IERC4626(vault).balanceOf(owner));
```
Comment (StErMi) [inline src/ERC4626Bundler.sol:112]: This is not correct. The maximum amount of redeemable tokens could be different from the owner's balance of vault's token. This is true for simpler vaults (like `sDAI` where `maxRedeem == user's balance`) but it could not be true for other vaults (timelock mechanism, withdrawal limits for any reason and so on).

Comment (Rubilmax) [reply src/ERC4626Bundler.sol:112]: Indeed, but we've chosen to not cover this kind of behavior onchain and rather rely on the offchain input to not exceed the owner's max redeemable shares

**Comment by StErMi** [comment]
Removing those utility checks allows the tx to fail when the op amount is above those thresholds (it will revert down the underlying vault flow).

At this point, I'm wondering what's the utility of leaving the other... 

Why having `shares = Math.min(shares, IERC4626(vault).balanceOf(owner));` (in redeem) but removing `assets = Math.min(assets, IERC4626(vault).maxWithdraw(owner));` in withdraw?

**Comment by Rubilmax** [comment]
We want to decrease the gas cost of depositing and withdrawing from MM vaults for which `maxDeposit` & `maxRedeem` are costly because they involve counting the assets and accrued interest managed by the vault

We consider standard use cases of ERC4626 to call `deposit` upon deposit and `redeem` upon withdraw, because it is natural for us to deposit a given amount of assets and get whatever amount of shares in exchange for them (modulo slippage protection); symmetrically it is natural for us to redeem a nb of shares (the ERC4626 token) in exchange for whatever amount of assets (modulo slippage)

We don't want to revert in case the bundler is some wei of underlying off due to some slippage at prior steps of execution of the bundle (e.g. stakeEth of stETH)
But we consider `maxDeposit` etc functions dummy for most vaults, including MM; and we are fine relying on the offchain to check there is a safety margin large enough to not revert inside the bundler

On the other hand, we don't have specific use cases for `mint` & `withdraw`. It seems that the case where the `min` would be useful inside these functions is whenever the bundler does not have enough assets to mint the given nb of shares, or when the bundler doesn't have enough shares to withdraw the given nb of assets
These use cases are inconsistent to us: in this case, the user should instead deposit the amount of assets or redeem the amount of shares

In other words: there is no clear use case where we'd like the bundler to NOT revert upon `mint` or `withdraw`, considering the inputs involved and the possibility to use `deposit` & `redeem` alternatively

File: src/ERC4626Bundler.sol (hunk 5 above)
Comment (StErMi) [reply src/ERC4626Bundler.sol:112]: I would at least suggest adding a comment because the behavior is different. You could end up not being able to withdraw all your balance (depending on the Vault's logic)

**Comment by StErMi** [comment]
Worst-case scenarios

- `erc4626Mint` reverts if the underlying vault has some limitation (global or on the `receiver`). For example, MM would revert if you can't deposit all the assets needed to mint those shares because the cap is lower of the transferred amount
- `erc4626Deposit` same scenario as `erc4626Mint`
- `erc4626Withdraw` reverts if the underlying vault has some limitation (global or on the `owner`). For example, the user has part of his/her balance time locked or for MM if there's not enough liquidity to withdraw the requested assets
- `erc4626Redeem` same scenario as `erc4626Withdraw` (but with share context)


From what I see, the change does not introduce any security problems per-se but could make the transaction reverts because it has removed some automatic upper bound checks that were, instead of reverting, limiting the `amount` that the user could deposit/mint or withdraw/redeem.
